ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/FutureAssertions.scala

```diff
@@ -34,13 +40,126 @@ final class FutureAssertions[T](future: Future[T]) {
+  )(
+      test: => Future[V]
+  )(implicit ec: ExecutionContext, logger: Logger): Future[V] = {
+    def internalSucceedsEventually(interval: FiniteDuration): Future[V] = {
+      val newMaxInterval = interval - delay
+      if (newMaxInterval < Duration.Zero) {
+        test.andThen { case Failure(exception) =>
+          logger.error(
+            s"Assertion never succeeded after $maxInterval with a delay of $delay. Description: $description",
+            exception,
+          )
+        }
+      } else {
+        if (newMaxInterval == Duration.Zero) {
+          assertAfter(delay)(test)
+        } else {
+          assertAfter(delay)(test).recoverWith { case NonFatal(ex) =>
+            logger.debug(
+              s"Failed assertion: $description. Running again with new max interval $newMaxInterval",
+              ex,
+            )
+            internalSucceedsEventually(newMaxInterval)
+          }
+        }
+      }
+    }
+
+    internalSucceedsEventually(maxInterval)
+  }
+
+  def succeedsUntil[V](
+      delay: FiniteDuration = 100.millis,
+      succeedDuration: FiniteDuration,
+      succeedDeadline: Option[Instant] = None,
+  )(
+      test: => Future[V]
+  )(implicit ec: ExecutionContext, logger: Logger): Future[V] = {
+    def internalSucceedsUntil(interval: FiniteDuration): Future[V] = {
+      val newMaxInterval = interval - delay
+      if (
+        succeedDeadline.exists(
+          _.isBefore(Instant.now().plusSeconds(delay.toSeconds))
+        ) || newMaxInterval < Duration.Zero
+      ) {
+        test
+      } else {
+        if (newMaxInterval == Duration.Zero) {
+          assertAfter(delay)(test)
+        } else {
+          assertAfter(delay)(test)
+            .flatMap { _ =>
+              internalSucceedsUntil(newMaxInterval)
+            }
+        }
+      }
+    }
+
+    internalSucceedsUntil(succeedDuration)
+      .andThen { case Failure(exception) =>
+        logger.error(
+          s"Repeated assertion failed with a succeed duration of $succeedDuration.",
+          exception,
+        )
+      }
+  }
+
+  def forAllParallel[T](
+      data: Seq[T]
+  )(testCase: T => Future[Unit])(implicit ec: ExecutionContext): Future[Seq[Unit]] = {
+    Future
+      .traverse(data)(input =>
+        testCase(input).map(Right(_)).recover { case NonFatal(ex) =>
+          Left(input -> ex)
+        }
+      )
+      .map { results =>
+        val (failures, successes) = results.partitionMap(identity)
+        if (failures.nonEmpty) {
+          throw ParallelTestFailureException(
+            s"Failed parallel test case. Failures: ${failures.length}. Success: ${successes.length}\nFailed inputs: ${failures
+              .map(_._1)
+              .mkString("[", ",", "]")}",
+            failures.last._2,
+          )
+        } else {
+          successes
+        }
+      }
+  }
+
+  def optionalAssertion(runs: Boolean, description: String)(
+      assertions: => Future[_]
+  )(logger: Logger): Future[_] = {
```

Comment:
Why not a contextualized logger defined in this file, and a logging context passed in?

Reply:
Also, nit: unneeded braces